crates/ruff_python_parser/src/parser/statement.rs

```diff
@@ -1443,40 +1432,51 @@ impl<'src> Parser<'src> {
         // Check if our assumption is incorrect and it's actually a parenthesized
         // expression.
         if !with_item_kind.is_parenthesized_expression() && self.at(TokenKind::Rpar) {
-            if self.peek() == TokenKind::Colon {
+            if has_optional_vars {
+                // If any of the with item has optional variables, then our assumption is
+                // correct and it is a parenthesized with items. Now, we need to restrict
+                // the grammar for a with item's context expression which is:
+                //
+                //     with_item: expression ...
+                //
+                // So, named, starred and yield expressions not allowed.
+                for parsed_with_item in &parsed_with_items {
+                    // Parentheses resets the precedence.
+                    if parsed_with_item.is_parenthesized {
+                        continue;
+                    }
+                    let err = match parsed_with_item.item.context_expr {
+                        Expr::Named(_) => ParseErrorType::UnparenthesizedNamedExpression,
+                        Expr::Starred(_) => ParseErrorType::StarredExpressionUsage,
+                        Expr::Yield(_) | Expr::YieldFrom(_) => {
+                            ParseErrorType::InvalidYieldExpressionUsage
+                        }
+                        _ => continue,
+                    };
+                    self.add_error(err, &parsed_with_item.item.context_expr);
+                }
```

Comment:
One or more with item have an optional variable, so it's a parenthesized with items. Limit the parsed expression by reporting error for invalid expressions in this context.